### PR TITLE
Cafe stuff!

### PR DESCRIPTION
**Pull request by NigglyWiggly**
Finally actually got around to doing the code for this, it's probably
horribly wrong somewhere, but it's worked just fine every time I tried
it.

### CODE REVIEW

**Comment by Arthurdentist** [comment]
AAAAAAH
MERGE CONFLICTS #3560

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 1]
```diff
@@ -264,6 +264,46 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+	var/obj/item/weapon/reagent_containers/glass/kettle = null
+	dispensable_reagents = list("tea","greentea","redtea", "coffee","milk","cream","water","hot_coco",
+	"soymilk")
+
+
+/obj/machinery/chem_dispenser/brewer/mapping
+	max_energy = 100
+	energy = 100
+
+/obj/machinery/chem_dispenser/brewer/togglePanelOpen(var/obj/toggleitem, mob/user)
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: You don't need this, it's covered in the parent.

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 2]
```diff
@@ -264,6 +264,46 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+	if(beaker)
+		user << "You can't reach the maintenance panel with a kettle in the way!"
+		return
+	return ..()
+
+/obj/machinery/chem_dispenser/brewer/attackby(var/obj/item/weapon/D as obj, var/mob/user as mob) //to be worked on
+
+
+	if(istype(D, /obj/item/weapon/reagent_containers/glass/kettle))
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: Is there any reason you can only use kettles in this machine?

File: code/modules/reagents/Chemistry-Machinery.dm [hunk 3]
```diff
@@ -264,6 +264,46 @@ USE THIS CHEMISTRY DISPENSER FOR MAPS SO THEY START AT 100 ENERGY
...
+	max_energy = 100
+	energy = 100
+
+/obj/machinery/chem_dispenser/brewer/togglePanelOpen(var/obj/toggleitem, mob/user)
+	if(beaker)
+		user << "You can't reach the maintenance panel with a kettle in the way!"
+		return
+	return ..()
+
+/obj/machinery/chem_dispenser/brewer/attackby(var/obj/item/weapon/D as obj, var/mob/user as mob) //to be worked on
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Machinery.dm]: You should inherit the parent of this somewhere.
Try 
`if(..())`
`    return 1`
at the start of the proc.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 4]
```diff
@@ -3350,6 +3351,7 @@
 	color = "#664300" // rgb: 102, 67, 0
 	adj_sleepy = 0
 	adj_temp = 5
+*/
 
 /datum/reagent/drink/coffee/soy_latte/on_mob_life(var/mob/living/M as mob)
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: Can't comment out soy_latte and leave this in.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 5]
```diff
@@ -3336,6 +3336,7 @@
 
 		holder.remove_reagent(src.id, 0.1)
 
+/*
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: Why comment this out?

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 6]
```diff
@@ -4467,3 +4469,275 @@
...
+	id = "chifir"
+	description = "Strong Russian tea, it'll help you remember what you had for lunch!"
+
+/datum/reagent/chifir/on_mob_life(var/mob/living/M as mob)
+	if(!M) M = holder.my_atom
+
+	if(ishuman(M) && prob(5))
+		var/mob/living/carbon/human/H=M
+		H.vomit()
+		holder.remove_reagent("chifir",volume) // Remove all charcoal.
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: Might as well do id, volume.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 7]
```diff
@@ -4467,3 +4469,275 @@
 	if(istype(holder))
 		holder.reagent_list -= src
 		holder = null
+
+
+
+//Tea and Coffee, hopefully
+
+
+/datum/reagent/drink/tea
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: The tea reagent already exists in the code.

File: code/modules/reagents/reagent_containers/food/drinks.dm [hunk 8]
```diff
@@ -727,6 +727,17 @@
 	icon_state = "britcup"
 	volume = 30
 
+/obj/item/weapon/reagent_containers/food/drinks/americup
+	name = "cup"
+	desc = "A cup with the American flag emblazoned on it."
+	icon_state = "americup"
+	volume = 30
+
+/obj/item/weapon/reagent_containers/food/drinks/empty
```
Comment (ComicIronic) [inline code/modules/reagents/reagent_containers/food/drinks.dm]: If this is a mug, why doesn't it come under the mug type?

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 9]
```diff
@@ -4467,3 +4469,275 @@
...
+	"blood"
+)
+
+
+/datum/reagent/drinks/tea/chifir
+	name = "Chifir"
+	id = "chifir"
+	description = "Strong Russian tea, it'll help you remember what you had for lunch!"
+
+/datum/reagent/chifir/on_mob_life(var/mob/living/M as mob)
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: This should be /datum/reagent/drinks/tea/chifir/on_mob_life

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 10]
```diff
@@ -4467,3 +4469,275 @@
...
+	id = "singularitea"
+	description = "Swirly!"
+
+var/global/list/chifir_doesnt_remove=list(
+	"chifir",
+	"blood"
+)
+
+
+/datum/reagent/drinks/tea/chifir
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: Should be /drink/ instead of /drinks/

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 11]
```diff
@@ -4467,3 +4469,275 @@
...
+
+
+/datum/reagent/drink/coffee/tonio
+	name = "Tonio"
+	id = "tonio"
+	nutriment_factor = 1 * FOOD_METABOLISM
+
+
+
+/datum/reagent/tonio/on_mob_life(var/mob/living/M as mob)
```
Comment (ComicIronic) [inline code/modules/reagents/Chemistry-Reagents.dm]: See the comment above as to pathing.

If this is an effect also present in charcoal and chifir, you should make it a general proc.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 12]
```diff
@@ -4467,3 +4468,201 @@
...
+
+	M.adjustToxLoss(-2*REM)
+	..()
+	return
+
+	if(!holder) return
+	M:nutrition += nutriment_factor
+	holder.remove_reagent(src.id, FOOD_METABOLISM)
+	if(!M) M = holder.my_atom
+	if(M:getOxyLoss() && prob(50)) M:adjustOxyLoss(-2)
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Reagents.dm]: COLONS

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 13]
```diff
@@ -4467,3 +4468,201 @@
...
+	if(M:getBruteLoss() && prob(60)) M:heal_organ_damage(2,0)
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Reagents.dm]: COLONS

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 14]
```diff
@@ -4467,3 +4468,201 @@
...
+	if(M:getFireLoss() && prob(50)) M:heal_organ_damage(0,2)
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Reagents.dm]: COLONS

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 15]
```diff
@@ -4467,3 +4468,201 @@
...
+	if(M:getToxLoss() && prob(50)) M:adjustToxLoss(-2)
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Reagents.dm]: COLONS

File: code/modules/reagents/reagent_containers/food/drinks/drinkingglass.dm [hunk 16]
```diff
@@ -483,3 +483,140 @@
...
+	name = "mug"
+	desc = "A simple mug."
+	icon = 'icons/obj/cafe.dmi'
+	icon_state = "mug_empty"
+	isGlass = 0
+	amount_per_transfer_from_this = 10
+	volume = 50
+	g_amt = 500
+
+	on_reagent_change()
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/food/drinks/drinkingglass.dm]: RELATIVE

File: code/modules/reagents/reagent_containers/glass.dm [hunk 17]
```diff
@@ -377,3 +377,53 @@
 		reagents.add_reagent("fluorosurfactant", 20)
 
 */
+
+/obj/item/weapon/reagent_containers/glass/kettle
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: what is Object Oriented Code

File: code/modules/reagents/reagent_containers/glass.dm [hunk 18]
```diff
@@ -377,3 +377,53 @@
...
+	name = "Kettle"
+	desc = "A pot designed to hold hot drinks. Can hold up to 50 units."
+	icon_state = "kettle"
+	m_amt = 200
+	volume = 50
+	w_type = RECYK_GLASS
+	amount_per_transfer_from_this = 10
+	flags = FPRINT  | OPENCONTAINER
+
+/obj/item/weapon/reagent_containers/glass/kettle/red
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: what is Object Oriented Code

File: code/modules/reagents/reagent_containers/glass.dm [hunk 19]
```diff
@@ -377,3 +377,53 @@
...
+	name = "Kettle"
+	desc = "A pot designed to hold hot drinks. Can hold up to 50 units."
+	icon_state = "kettle_red"
+	m_amt = 200
+	volume = 50
+	w_type = RECYK_GLASS
+	amount_per_transfer_from_this = 10
+	flags = FPRINT  | OPENCONTAINER
+
+/obj/item/weapon/reagent_containers/glass/kettle/blue
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: what is Object Oriented Code

File: code/modules/reagents/reagent_containers/glass.dm [hunk 20]
```diff
@@ -377,3 +377,53 @@
...
+	name = "Kettle"
+	desc = "A pot designed to hold hot drinks. Can hold up to 50 units."
+	icon_state = "kettle_blue"
+	m_amt = 200
+	volume = 50
+	w_type = RECYK_GLASS
+	amount_per_transfer_from_this = 10
+	flags = FPRINT  | OPENCONTAINER
+
+/obj/item/weapon/reagent_containers/glass/kettle/purple
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: what is Object Oriented Code

File: code/modules/reagents/reagent_containers/glass.dm [hunk 21]
```diff
@@ -377,3 +377,53 @@
...
+	name = "Kettle"
+	desc = "A pot designed to hold hot drinks. Can hold up to 50 units."
+	icon_state = "kettle_purple"
+	m_amt = 200
+	volume = 50
+	w_type = RECYK_GLASS
+	amount_per_transfer_from_this = 10
+	flags = FPRINT  | OPENCONTAINER
+
+/obj/item/weapon/reagent_containers/glass/kettle/green
```
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: what is Object Oriented Code

File: code/modules/reagents/Chemistry-Recipes.dm [hunk 22]
```diff
@@ -2414,3 +2413,122 @@ datum
...
+
+
+		cappuccino
+			name = "Cappuccino"
+			id = "cappuccino"
+			result = "cappuccino"
+			required_reagents = list("espresso" = 1, "cream" = 1)
+			result_amount = 2
+
+		tonio
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Recipes.dm]: are you fucking kidding me

this is literally doctor's delight: improved edition

fuck off with this shit

**Comment by NigglyWiggly** [comment]
>this is literally doctor's delight: improved edition

>fuck off with this shit

Oh I'm sorry, would you have preferred if I called it 'Omnizine'? :^))

>COLONS
>COLONS
>COLONS
>COLONS

You know that's just copy-pasted from DD right

**Comment by Iamgoofball** [comment]
@NigglyWiggly at least omnizine has downsides, dumbass. this has none

also copypaste is not an excuse for this bullshit

copypaste is fucking retarded too, stop being lazy and fix it

**Comment by NigglyWiggly** [comment]
>copypaste is fucking retarded too, stop being lazy and fix it

You really overestimate my ability here. I'm honestly surprised the whole thing works at all.

**Comment by clusterfack** [comment]
To go more indepth

Colons as I understand it are only ever to be used when you want to assert something is true without making a variable of a specific type (like it has this specific proc or variable name), we don't use it in our code.

The kettle is a problem because all your kettle types inherit from kettle. So they will have all the same variables and values if you give the values to kettle, then you only need to put in the types like kettle/red the new color.

Procs can be defined in byond by just having a tab, thats called relative pathing. We don't use that either because it makes things unreadable and unsearchable. When defining procs for something use the fullobjectpathname/procname() for a proc that is defined earlier on a parent, or if it is a brand new proc specifically for that object fullobjectpathname/proc/procname().

**Comment by Boxe** [comment]
fix it nerd

**Comment by Iamgoofball** [comment]
this pull request appears to be fucked

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 23]
```diff
@@ -4467,3 +4467,195 @@
...
+		if(reagent.id in tonio_doesnt_remove)
+			continue
+		holder.remove_reagent(reagent.id, 3*REM)
+
+	M.adjustToxLoss(-2*REM)
+	..()
+	return
+
+	if(!holder) return
+	M:nutrition += nutriment_factor
```
Comment (Iamgoofball) [inline code/modules/reagents/Chemistry-Reagents.dm]: colons

**Comment by NigglyWiggly** [comment]
Bite the pillow

**Comment by d3athrow** [comment]
>goofball calling people out on copypasting and overpowered chems

Hypocrisy

**Comment by Iamgoofball** [comment]
@d3athrow at least I don't do colons

File: code/modules/reagents/reagent_containers/glass.dm (hunk 18 above)
Comment (ComicIronic) [inline code/modules/reagents/reagent_containers/glass.dm]: Literally this.
How would you implement it without varediting?

**Comment by Senaat1** [comment]
Goofball, i don't mean to be a cunt, but kill yourself my man. You are literally worse than Duny, i don't know much about coding but from a player point of view all of your features except for the fucking kitchen additions seem like absolute garbage, especially mining. Don't shittalk people when you are this fucking worthless. I will go out of my way to make sure Goofchem never gets merged.

**Comment by 9600bauds** [comment]
"Except" for the kitchen additions?

**Comment by NigglyWiggly** [comment]
I'm just gonna declare this PR unsalvagable because I'm terrible at everything

**Comment by Probe1** [comment]
shutup hime you're wonderful and I'm looking forward to it when you do merge it.

File: code/modules/reagents/reagent_containers/glass.dm (hunk 18 above)
Comment (Iamgoofball) [inline code/modules/reagents/reagent_containers/glass.dm]: outdated comment from when it wasn't OOP

**Pull request closed without merging**